src/vmm/src/lib.rs

```diff
@@ -393,6 +393,15 @@ impl Vmm {
 
         Vcpu::register_kick_signal_handler();
 
+        #[cfg(target_arch = "x86_64")]
+        if let Some(devices::BusDevice::CpuContainer(cont)) =
+            self.get_bus_device(DeviceType::CpuContainer, "CpuContainer")
+        {
+            let mut locked_container = cont.lock().expect("Poisoned lock");
+            for i in self.vcpus_handles.len()..(self.vcpus_handles.len() + vcpu_count) {
+                locked_container.cpu_devices[i].inserting = true;
+            }
+        }
```

Comment:
Mhh, is this correct to do for the boot cpus? I guess when we construct the boot CPUs, we haven't attached the cpucontainer to the bus yet, and thus we don't set the "inserting" bit, but that sounds quite brittle. Can we explicitly prevent this from happening on the boot cpus?

Reply:
Hmm, I've been thinking about the possible effects of this even if the cpucontainer is attached. What will happen is that when `CSCN` AML method is called, AML will notify the device with 0x1, device check. From ACPI spec:

> Device Check. Used to notify OSPM that the device either appeared or disappeared. If the device has appeared, OSPM will re-enumerate from the parent. If the device has disappeared, OSPM will invalidate the state of the device. OSPM may optimize out re-enumeration.

So this may cause a reenumeration of the Processor container object `PNP0A05`/`ACPI0010`. This may also be optimised out, I can't tell. But I think it's probably best not to send the notification out anyway. There is also possibility that this could cause initiate an undock, if a `_DCK` object is ever present within the `CpuDevice` AML (not sure it would be, but it's a possibility I suppose). From the same paragraph of the spec:

> If _DCK is present, then Notify(object,1) is assumed to indicate an undock request. 

For this reason, I think it's best to perhaps move this block out of `start_vcpus` and put it into the `hotplug_vcpus` function, so that `inserting` is only set for the newly added vCPUs. I will just need to set `CpuDevices` as `active` differently, but this should be easy enough to do, just pass in the boot `cpu_count` as a parameter to the CpuContainer constructor. That should be fairly easy.

Reply:
Update: I've made the changes I mentioned.